File: testdata/fixtures/go/CWE-918-2.go
package fixture

import (
	"net/http"
)

func fetchesWhateverItIsAsked(w http.ResponseWriter, r *http.Request) {
	target := r.URL.Query().Get("url")
	// ruleid: go-ssrf
	resp, _ := http.Get(target)
	_ = resp
}

func postsToAnyHost(r *http.Request) error {
	hook := r.FormValue("callback")
	// ruleid: go-ssrf
	req, _ := http.NewRequest("POST", hook, nil)
	_ = req
	return nil
}

func clientGet(r *http.Request, client *http.Client) error {
	url := r.Header.Get("X-URL")
	// ruleid: go-ssrf
	resp, _ := client.Get(url)
	_ = resp
	return nil
}

func doAnyHost(r *http.Request, client *http.Client) error {
	target := r.URL.Query().Get("target")
	// ruleid: go-ssrf
	req, _ := http.NewRequest("GET", target, nil)
	_ = req
	return nil
}

// ok: go-ssrf
func fetchesFixedEndpoint() error {
	resp, _ := http.Get("https://api.example.com/v1/status")
	_ = resp
	return nil
}

// ok: go-ssrf
func fetchesConfigured(url string) error {
	resp, _ := http.Get(url)
	_ = resp
	return nil
}

File: testdata/fixtures/java/CWE-918-2.java
package fixture;

import java.io.InputStream;
import java.net.HttpURLConnection;
import java.net.URL;
import java.net.URI;
import javax.servlet.http.HttpServletRequest;
import org.apache.http.client.methods.HttpGet;

class Cwe918b {

    InputStream viaUrl(HttpServletRequest request) throws Exception {
        String target = request.getParameter("url");
        // ruleid: java-ssrf
        URL url = new URL(target);
        HttpURLConnection conn = (HttpURLConnection) url.openConnection();
        return conn.getInputStream();
    }

    HttpGet viaHttpGet(HttpServletRequest request) {
        String where = request.getParameter("callback");
        // ruleid: java-ssrf
        return new HttpGet(where);
    }

    URI viaUriCreate(HttpServletRequest request) {
        String host = request.getParameter("host");
        String full = "http://" + host + "/status";
        // ruleid: java-ssrf
        return URI.create(full);
    }

    InputStream viaHeader(HttpServletRequest request) throws Exception {
        String upstream = request.getHeader("X-Target");
        // ruleid: java-ssrf
        URL url = new URL(upstream);
        return url.openStream();
    }

    // ok: java-ssrf
    InputStream safeConstant() throws Exception {
        URL url = new URL("https://api.example.com/status");
        return url.openStream();
    }

    // ok: java-ssrf
    InputStream safeFromConfig(String cfg) throws Exception {
        URL url = new URL(cfg);
        return url.openStream();
    }

    void safeWithFixedHost(HttpServletRequest request) throws Exception {
        String page = request.getParameter("page");
        // ok: java-ssrf - path under fixed host not host itself
        String fixed = "https://intranet.example.com/static";
        new URL(fixed);
    }
}
